toolchain/check/testdata/primitives/string_literals.carbon

```diff
@@ -0,0 +1,171 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/uint.carbon
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/primitives/string_literals.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/primitives/string_literals.carbon
+
+// --- fail_no_string.carbon
+
+library "[[@TEST_NAME]]";
+
+fn F() {
+  // CHECK:STDERR: fail_no_string.carbon:[[@LINE+4]]:3: error: name `Core.String` implicitly referenced here, but not found [CoreNameNotFound]
+  // CHECK:STDERR:   "hello";
+  // CHECK:STDERR:   ^~~~~~~
+  // CHECK:STDERR:
+  "hello";
+}
+
+// --- unusual_string.carbon
+
+package Core library "[[@TEST_NAME]]";
+
+import library "prelude/parts/uint";
+
+class String {
+  var begin: u8*;
+  var end: u8*;
+}
+
+// --- fail_use_unusual_string.carbon
+
+import Core library "unusual_string";
+
+// CHECK:STDERR: fail_use_unusual_string.carbon:[[@LINE+4]]:14: error: unexpected representation for type `Core.String` [StringLiteralTypeUnexpected]
+// CHECK:STDERR: let x: str = "hello";
+// CHECK:STDERR:              ^~~~~~~
+// CHECK:STDERR:
+let x: str = "hello";
+
+// --- incomplete_string.carbon
+
+package Core library "[[@TEST_NAME]]";
+
+class String;
+
+// --- fail_use_incomplete_string.carbon
+
+library "[[@TEST_NAME]]";
+
+import Core library "incomplete_string";
+
+fn F() {
+  // CHECK:STDERR: fail_use_incomplete_string.carbon:[[@LINE+7]]:3: error: type `Core.String` is incomplete [StringLiteralTypeIncomplete]
+  // CHECK:STDERR:   "hello";
+  // CHECK:STDERR:   ^~~~~~~
+  // CHECK:STDERR: incomplete_string.carbon:4:1: note: class was forward declared here [ClassForwardDeclaredHere]
+  // CHECK:STDERR: class String;
+  // CHECK:STDERR: ^~~~~~~~~~~~~
+  // CHECK:STDERR:
+  "hello";
+}
+
+// --- small_string.carbon
+
+package Core library "[[@TEST_NAME]]";
+
+import library "prelude/parts/uint";
+
+class String {
+  var ptr: u8*;
+  var size: u8;
```

Comment:
Maybe a comment to help the reader understand what is small and how it'll be used later on?

Reply:
Added.